packages/flutter/lib/src/foundation/consolidate_response.dart

```diff
@@ -3,31 +3,103 @@
 // found in the LICENSE file.
 
 import 'dart:async';
+import 'dart:convert';
 import 'dart:io';
 import 'dart:typed_data';
 
-/// Efficiently converts the response body of an [HttpClientResponse] into a [Uint8List].
+/// Signature for getting notified when chunks of bytes are received while
+/// consolidating the bytes of an [HttpClientResponse] into a [Uint8List].
+///
+/// The `cumulative` parameter will contain the total number of bytes received
+/// thus far. If the response has been gzipped, this number will be the number
+/// of compressed bytes that have received _across the wire_.
+///
+/// The `total` parameter will contain the _expected_ total number of bytes to
+/// be received across the wire (extracted from the value of the
+/// `Content-Length` HTTP response header), or -1 if the size of the response
+/// body is not known in advance.
+///
+/// This is used in [consolidateHttpClientResponseBytes].
+typedef BytesReceivedCallback = void Function(int cumulative, int total);
+
+/// Efficiently converts the response body of an [HttpClientResponse] into a
+/// [Uint8List].
 ///
 /// The future returned will forward all errors emitted by [response].
-Future<Uint8List> consolidateHttpClientResponseBytes(HttpClientResponse response) {
-  // response.contentLength is not trustworthy when GZIP is involved
-  // or other cases where an intermediate transformer has been applied
-  // to the stream.
+///
+/// The [onBytesReceived] callback, if specified, will be invoked for every
+/// chunk of bytes that are received while consolidating the response bytes.
+/// For more information on how to interpret the parameters to the callback,
+/// see the documentation on [BytesReceivedCallback].
+///
+/// If the [response] is gzipped, this will automatically un-compress the
+/// bytes in the returned list (even if [HttpClient.autoUncompress] is false).
+// TODO(tvolkert): Remove the [client] param once https://github.com/dart-lang/sdk/issues/36971 is fixed.
+Future<Uint8List> consolidateHttpClientResponseBytes(
+  HttpClientResponse response, {
+  HttpClient client,
+  BytesReceivedCallback onBytesReceived,
+}) {
   final Completer<Uint8List> completer = Completer<Uint8List>.sync();
-  final List<List<int>> chunks = <List<int>>[];
-  int contentLength = 0;
+
+  final _OutputBuffer output = _OutputBuffer();
+  ByteConversionSink sink = output;
+  int expectedContentLength = response.contentLength;
+  if (response.headers?.value(HttpHeaders.contentEncodingHeader) == 'gzip') {
```

Comment:
what if it's "identity, gzip"?

Reply:
I wondered about making this more robust, but I opted for copying the exact behavior from `HttpClient` so as to not add confusion as to how we determine that unzipping is necessary.

Reply:
Please at a minimum file a bug and add a TODO

Reply:
According to [the spec](https://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html#sec3.5), "identity" is an invalid value in the Content-Encoding header.

> identity The default (identity) encoding; the use of no transformation whatsoever. This content-coding is used only in the Accept- Encoding header, and SHOULD NOT be used in the Content-Encoding header.

So I think the current impl here is correct.